Unified named colors in RichTextLabel

dalexeev opened the pull request:
Now the BB tag `[color]` uses the color names from the `Color` class.
Also, the use of the `#` symbol in an HTML color code has become optional.

---

Please note that there is backward incompatibility with several old color names.

![200124-1](https://user-images.githubusercontent.com/47700418/73067885-db29a480-3eba-11ea-85c7-a1865fd68c33.png)

dankan1890 commented:
Nice copy/paste.

bojidar-bg commented:
Maybe, to lessen the compatibility breakage, `grey` and `navy` could be added to the list of Colors in Color?

dalexeev commented:
Should the `grey` color be `#bfbfbf` or `#808080`? If `gray` and `grey` are different, it will be strange. If `grey` is `#bfbfbf`, then compatibility will still be broken.

bojidar-bg commented:
Well, hm, didn't consider that. Feel free to leave it as is, in this case.

dalexeev commented:
I watched https://en.wikipedia.org/wiki/X11_color_names and noticed the 'Alternatives' column in the colors table:
![200520-1](https://user-images.githubusercontent.com/47700418/82416593-f4b2b580-9a82-11ea-9674-83265019277e.png)

I'm not sure if we really need this, but in any case, this is a matter of a separate PR.

dalexeev commented:
Maybe, since similar code is used in several places, do we need to add the `Color::html_or_named` function or something like that?

https://github.com/godotengine/godot/blob/022c615490afead6bf34819fd2b3bf8cb53c5900/scene/gui/rich_text_label.cpp#L1452-L1459

https://github.com/godotengine/godot/blob/022c615490afead6bf34819fd2b3bf8cb53c5900/editor/editor_help.cpp#L1378-L1386

dalexeev committed:
Unified named colors in RichTextLabel

Now the BB tag `[color]` uses the color names from the `Color` class.
Also, the use of the `#` symbol in an HTML color code has become optional.

dalexeev commented:
Rebased and ready to merge.

aaronfranke reviewed (approved):
I tested this, throwing everything I could think of at it (3, 4, 6, 8 digit hex colors, with and without `#`, and named), and it works perfectly.

Additionally, I love the implementation. I believe the idea was to replace `ColorN` and from HTML with a constructor that accepts a string, and `Color::from_string` works *perfectly* for this. Although that can be done in another PR, since it would also involve removing `ColorN` which would be a big change.

dalexeev commented:
To be honest, it seems to me too that we have too many color constructors from a string: `Color::html`, `Color::named`, and now also `Color::from_string`. In GDScript, `Color(string)` uses `Color::html`.

![](https://user-images.githubusercontent.com/47700418/101295157-ef556680-382c-11eb-802c-039b6ea65bac.png)

It seems to me that it would be more appropriate if in GDScript `Color(string)` would use `Color::from_string` instead. In this case, it will be possible to use both `Color("#ff0000")` and `Color("red")`. Then `ColorN("red ")` won't really be needed.

Also, I had a question if the methods `find_named_color`, `get_named_color_count`, `get_named_color_name`, `get_named_color` should be made private, but I decided not to touch that. First of all, I wanted to fix the inconsistency in `RichTextLabel`, and I made changes to the core only when necessary. For other changes, we probably need a separate PR.

akien-mga commented:
> It seems to me that it would be more appropriate if in GDScript `Color(string)` would use `Color::from_string` instead. In this case, it will be possible to use both `Color("#ff0000")` and `Color("red")`. Then `ColorN("red")` won't really be needed.

That's what I planned to implement after a recent discussion with @reduz.

akien-mga commented:
Thanks!

akien-mga commented:
Welp, should have rebased before merging as there's some more uses of `_get_color_from_string` to convert.
